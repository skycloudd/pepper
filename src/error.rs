use crate::ir::FileId;
use chumsky::{
    error::{Rich, RichReason},
    span::SimpleSpan,
};
use core::ops::Range;

#[derive(Clone, Debug)]
pub enum Error {
    ExpectedFound {
        expected: Vec<String>,
        found: Option<String>,
        span: Range<usize>,
    },
    Custom {
        message: String,
        span: Range<usize>,
    },
    Many(Vec<Error>),
}

fn convert_inner(reason: &RichReason<String, &str>, span: Range<usize>) -> Error {
    match reason {
        RichReason::ExpectedFound { expected, found } => Error::ExpectedFound {
            expected: expected.iter().map(ToString::to_string).collect(),
            found: found.as_ref().map(|f| f.to_string()),
            span,
        },
        RichReason::Custom(message) => Error::Custom {
            message: message.clone(),
            span,
        },
        RichReason::Many(reasons) => Error::Many(
            reasons
                .iter()
                .map(|r| convert_inner(r, span.clone()))
                .collect(),
        ),
    }
}

impl From<Rich<'_, String, SimpleSpan<usize, FileId<'_>>, &str>> for Error {
    fn from(error: Rich<String, SimpleSpan<usize, FileId<'_>>, &str>) -> Self {
        convert_inner(error.reason(), error.span().start..error.span().end)
    }
}

impl From<&Rich<'_, String, SimpleSpan<usize, FileId<'_>>, &str>> for Error {
    fn from(error: &Rich<String, SimpleSpan<usize, FileId<'_>>, &str>) -> Self {
        convert_inner(error.reason(), error.span().start..error.span().end)
    }
}
