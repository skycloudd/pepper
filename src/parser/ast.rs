use crate::{lexer::tokens::Interned, span::Spanned};

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Ast(pub Vec<Spanned<Item>>);

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Item {
    Function(Spanned<Function>),
    Import(Spanned<Path>),
    Struct(Spanned<Struct>),
    Enum(Spanned<Enum>),
    Module(Spanned<Module>),
}

#[allow(clippy::module_name_repetitions)]
pub type AstType = Type<Path>;

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Path {
    pub base: Spanned<Interned>,
    pub segments: Vec<Spanned<Interned>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Module {
    File(Spanned<Interned>),
    Submodule {
        name: Spanned<Interned>,
        ast: Spanned<Ast>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Function {
    pub name: Spanned<Interned>,
    pub params: Spanned<Vec<Spanned<FunctionParam>>>,
    pub return_ty: Option<Spanned<AstType>>,
    pub body: Spanned<Block>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct FunctionParam {
    pub name: Spanned<Interned>,
    pub ty: Spanned<AstType>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Struct {
    pub name: Spanned<Interned>,
    pub fields: Spanned<Vec<Spanned<StructField>>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct StructField {
    pub name: Spanned<Interned>,
    pub ty: Spanned<AstType>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Enum {
    pub name: Spanned<Interned>,
    pub variants: Spanned<Vec<Spanned<EnumVariant>>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum EnumVariant {
    Unit(Spanned<Interned>),
    Tuple {
        name: Spanned<Interned>,
        fields: Spanned<Vec<Spanned<AstType>>>,
    },
    Struct {
        name: Spanned<Interned>,
        fields: Spanned<Vec<Spanned<StructField>>>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Statement {
    Expression(Spanned<Expression>),
    VarDecl {
        name: Spanned<Interned>,
        ty: Option<Spanned<AstType>>,
        value: Spanned<Expression>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Expression {
    Int(Spanned<Interned>),
    Float(Spanned<Interned>),
    Bool(Spanned<Interned>),
    String(Spanned<Interned>),
    Name(Spanned<Path>),
    Block(Spanned<Box<Block>>),
    Tuple(Spanned<Vec<Spanned<Self>>>),
    List(Spanned<Vec<Spanned<Self>>>),
    For {
        pattern: Spanned<Box<Pattern>>,
        iter: Spanned<Box<Self>>,
        body: Spanned<Box<Block>>,
    },
    BinaryOp {
        op: Spanned<BinaryOp>,
        lhs: Spanned<Box<Self>>,
        rhs: Spanned<Box<Self>>,
    },
    UnaryOp {
        op: Spanned<UnaryOp>,
        expr: Spanned<Box<Self>>,
    },
    Call {
        callee: Spanned<Box<Self>>,
        args: Spanned<Vec<Spanned<Self>>>,
    },
    Match {
        expr: Spanned<Box<Self>>,
        arms: Spanned<Vec<Spanned<MatchArm>>>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Block {
    pub statements: Vec<Spanned<Statement>>,
    pub return_expr: Option<Spanned<Expression>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct MatchArm {
    pub pattern: Spanned<Pattern>,
    pub body: Spanned<Expression>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Pattern {
    pub pattern_type: Spanned<PatternType>,
    pub condition: Option<Spanned<Expression>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum PatternType {
    Name(Spanned<Path>),
    Int(Spanned<Interned>),
    Float(Spanned<Interned>),
    Bool(Spanned<Interned>),
    String(Spanned<Interned>),
    Tuple(Spanned<Vec<Spanned<Pattern>>>),
    List(Spanned<Vec<Spanned<ListPattern>>>),
    TupleType {
        name: Spanned<Path>,
        fields: Spanned<Vec<Spanned<Pattern>>>,
    },
    StructType {
        name: Spanned<Path>,
        fields: Spanned<Vec<Spanned<StructPatternField>>>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct StructPatternField {
    pub name: Spanned<Interned>,
    pub pattern: Option<Spanned<Pattern>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum ListPattern {
    Pattern(Spanned<Pattern>),
    Rest,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    LessEquals,
    GreaterEquals,
    Less,
    Greater,
    Equals,
    NotEquals,
}

impl core::fmt::Display for BinaryOp {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Add => write!(f, "+"),
            Self::Sub => write!(f, "-"),
            Self::Mul => write!(f, "*"),
            Self::Div => write!(f, "/"),
            Self::LessEquals => write!(f, "<="),
            Self::GreaterEquals => write!(f, ">="),
            Self::Less => write!(f, "<"),
            Self::Greater => write!(f, ">"),
            Self::Equals => write!(f, "=="),
            Self::NotEquals => write!(f, "!="),
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum UnaryOp {
    Neg,
    Not,
}

impl core::fmt::Display for UnaryOp {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Neg => write!(f, "-"),
            Self::Not => write!(f, "!"),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Type<P> {
    Primitive(Spanned<P>),
    Tuple(Spanned<Vec<Spanned<Self>>>),
    Never,
    Function {
        params: Spanned<Vec<Spanned<Self>>>,
        return_ty: Spanned<Box<Self>>,
    },
}

impl<P: core::fmt::Display> core::fmt::Display for Type<P> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Primitive(primitive) => write!(f, "{}", primitive.0),
            Self::Tuple(inner) => {
                write!(f, "(")?;
                for (i, ty) in inner.iter().enumerate() {
                    if i != 0 {
                        write!(f, ", ")?;
                    }
                    write!(f, "{}", ty.0)?;
                }
                write!(f, ")")
            }
            Self::Never => write!(f, "!"),
            Self::Function { params, return_ty } => {
                write!(f, "func (")?;
                for (i, param) in params.iter().enumerate() {
                    if i != 0 {
                        write!(f, ", ")?;
                    }
                    write!(f, "{}", param.0)?;
                }
                write!(f, ") -> {}", return_ty.0)
            }
        }
    }
}
