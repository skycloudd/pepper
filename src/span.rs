use core::num::NonZeroUsize;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Span {
    start: usize,
    end: usize,
    file_id: FileId,
}

impl Span {
    pub fn zero(file_id: FileId) -> Self {
        chumsky::span::Span::new(file_id, 0..0)
    }

    pub const fn range(&self) -> core::ops::Range<usize> {
        self.start..self.end
    }
}

impl chumsky::span::Span for Span {
    type Context = FileId;

    type Offset = usize;

    fn new(context: Self::Context, range: core::ops::Range<Self::Offset>) -> Self {
        Self {
            start: range.start,
            end: range.end,
            file_id: context,
        }
    }

    fn context(&self) -> Self::Context {
        self.file_id
    }

    fn start(&self) -> Self::Offset {
        self.start
    }

    fn end(&self) -> Self::Offset {
        self.end
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct FileId(NonZeroUsize);

impl FileId {
    pub fn new(id: usize) -> Option<Self> {
        Some(Self(NonZeroUsize::new(id.checked_add(1)?)?))
    }

    pub fn get(self) -> usize {
        usize::from(self.0) - 1
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Spanned<T>(pub T, Span);

impl<T> Spanned<T> {
    pub const fn new(value: T, span: Span) -> Self {
        Self(value, span)
    }

    pub const fn span(&self) -> Span {
        self.1
    }

    pub fn boxed(self) -> Spanned<Box<T>> {
        Spanned::new(Box::new(self.0), self.1)
    }

    pub fn map<U>(self, f: impl FnOnce(T) -> U) -> Spanned<U> {
        Spanned::new(f(self.0), self.1)
    }

    pub fn map_with_span<U>(self, f: impl FnOnce(T, Span) -> U) -> Spanned<U> {
        Spanned::new(f(self.0, self.1), self.1)
    }

    pub fn map_span(self, f: impl FnOnce(Span) -> Span) -> Self {
        Self(self.0, f(self.1))
    }

    pub const fn as_ref(&self) -> Spanned<&T> {
        Spanned::new(&self.0, self.1)
    }
}

impl<T, E> Spanned<Result<T, E>> {
    pub fn transpose(self) -> Result<Spanned<T>, E> {
        self.0.map(|value| Spanned::new(value, self.1))
    }
}

impl<T> Spanned<Box<T>> {
    pub fn unbox(self) -> Spanned<T> {
        Spanned::new(*self.0, self.1)
    }
}

impl<T: Clone> Spanned<&T> {
    pub fn cloned(self) -> Spanned<T> {
        self.map(Clone::clone)
    }
}

impl<T> core::ops::Deref for Spanned<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> core::ops::DerefMut for Spanned<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

#[cfg(test)]
impl<T: serde::Serialize> serde::Serialize for Spanned<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.0.serialize(serializer)
    }
}
