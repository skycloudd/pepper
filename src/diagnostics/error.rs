use super::{Diag, ErrorSpan};
use crate::span::Span;
use chumsky::error::{Rich, RichReason};
use codespan_reporting::diagnostic::Severity;
use core::fmt::Display;
use owo_colors::OwoColorize;
use std::borrow::Cow;

#[derive(Clone, Debug)]
pub enum Error {
    ExpectedFound {
        expected: Vec<String>,
        found: Option<String>,
        span: Span,
    },
    Custom {
        message: String,
        span: Span,
    },
    AmbiguousModule {
        module_name: &'static str,
        module_name_span: Span,
        filenames: Vec<String>,
    },
    MissingModule {
        module_name: &'static str,
        module_name_span: Span,
        filenames: Vec<String>,
    },
}

impl Diag for Error {
    fn message(&self) -> Cow<str> {
        match self {
            Self::ExpectedFound {
                expected,
                found,
                span: _,
            } => if expected.is_empty() {
                "Unexpected end of file".to_string()
            } else if expected.len() == 1 {
                format!(
                    "Expected {}, but found {}",
                    expected[0].green(),
                    found.as_deref().unwrap_or("something else")
                )
            } else {
                format!(
                    "Expected one of {}, but found {}",
                    expected
                        .iter()
                        .map(|expected| expected.green().to_string())
                        .collect::<Vec<_>>()
                        .join(", "),
                    found.as_deref().unwrap_or("something else")
                )
            }
            .into(),
            Self::Custom { message, span: _ } => message.into(),
            Self::AmbiguousModule {
                module_name: _,
                module_name_span: _,
                filenames: _,
            } => "Ambiguous module name".into(),
            Self::MissingModule {
                module_name: _,
                module_name_span: _,
                filenames: _,
            } => "Missing module".into(),
        }
    }

    fn spans(&self) -> Vec<ErrorSpan> {
        match self {
            Self::ExpectedFound {
                expected: _,
                found,
                span,
            } => found.as_ref().map_or_else(
                || vec![ErrorSpan::primary(*span)],
                |found| vec![ErrorSpan::primary_message(format!("Found {found}"), *span)],
            ),
            Self::Custom { message: _, span } => vec![ErrorSpan::primary(*span)],
            Self::AmbiguousModule {
                module_name: _,
                module_name_span,
                filenames: _,
            }
            | Self::MissingModule {
                module_name: _,
                module_name_span,
                filenames: _,
            } => vec![ErrorSpan::primary_message(
                "Declared here",
                *module_name_span,
            )],
        }
    }

    fn notes(&self) -> Vec<String> {
        match self {
            Self::AmbiguousModule {
                module_name,
                module_name_span: _,
                filenames,
            } => core::iter::once(format!(
                "There exist multiple possible files for the module `{module_name}`:"
            ))
            .chain(filenames.iter().map(|filename| format!("  - {filename}")))
            .collect(),
            Self::MissingModule {
                module_name,
                module_name_span: _,
                filenames,
            } => core::iter::once(format!(
                "Could not find a file for the module `{module_name}`"
            ))
            .chain(core::iter::once(
                "These are the possible files for this module:".into(),
            ))
            .chain(filenames.iter().map(|filename| format!("  - {filename}")))
            .collect(),
            Self::ExpectedFound { .. } | Self::Custom { .. } => {
                vec![]
            }
        }
    }

    fn kind(&self) -> Severity {
        Severity::Error
    }
}

pub fn convert(error: &Rich<impl Display, Span, &str>) -> Vec<Error> {
    fn convert_inner(reason: &RichReason<impl Display, &str>, span: Span) -> Vec<Error> {
        match reason {
            RichReason::ExpectedFound { expected, found } => vec![Error::ExpectedFound {
                expected: expected.iter().map(ToString::to_string).collect(),
                found: found.as_ref().map(|f| f.to_string()),
                span,
            }],
            RichReason::Custom(message) => vec![Error::Custom {
                message: message.to_owned(),
                span,
            }],
            RichReason::Many(reasons) => reasons
                .iter()
                .flat_map(|r| convert_inner(r, span))
                .collect(),
        }
    }

    convert_inner(error.reason(), *error.span())
}
