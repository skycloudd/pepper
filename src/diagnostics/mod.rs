use crate::span::Span;
use codespan_reporting::diagnostic::{LabelStyle, Severity};
use std::borrow::Cow;

pub mod error;
pub mod report;

pub trait Diag {
    fn message(&self) -> Cow<str>;
    fn spans(&self) -> Vec<ErrorSpan>;
    fn notes(&self) -> Vec<String>;
    fn kind(&self) -> Severity;
}

#[derive(Debug)]
pub struct ErrorSpan {
    message: Option<String>,
    span: Span,
    label_style: LabelStyle,
}

impl ErrorSpan {
    pub const fn primary(span: Span) -> Self {
        Self {
            message: None,
            span,
            label_style: LabelStyle::Primary,
        }
    }

    pub fn primary_message(message: impl Into<String>, span: Span) -> Self {
        Self {
            message: Some(message.into()),
            span,
            label_style: LabelStyle::Primary,
        }
    }

    pub const fn secondary(span: Span) -> Self {
        Self {
            message: None,
            span,
            label_style: LabelStyle::Secondary,
        }
    }

    pub fn secondary_message(message: impl Into<String>, span: Span) -> Self {
        Self {
            message: Some(message.into()),
            span,
            label_style: LabelStyle::Secondary,
        }
    }
}
