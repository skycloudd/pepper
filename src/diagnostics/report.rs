use super::{Diag, ErrorSpan};
use chumsky::span::Span as _;
use codespan_reporting::diagnostic::{Diagnostic, Label};

pub fn report(diagnostic: &impl Diag) -> Diagnostic<usize> {
    codespan_reporting::diagnostic::Diagnostic::<usize>::new(diagnostic.kind())
        .with_message(diagnostic.message())
        .with_labels(
            diagnostic
                .spans()
                .into_iter()
                .map(|span| match span {
                    ErrorSpan::Primary(message, span) => {
                        let mut label =
                            Label::primary(span.context().id(), span.start()..span.end());

                        if let Some(message) = message {
                            label = label.with_message(message);
                        }

                        label
                    }
                    ErrorSpan::Secondary(message, span) => {
                        let mut label =
                            Label::secondary(span.context().id(), span.start()..span.end());

                        if let Some(message) = message {
                            label = label.with_message(message);
                        }

                        label
                    }
                })
                .collect(),
        )
        .with_notes(diagnostic.notes())
}
