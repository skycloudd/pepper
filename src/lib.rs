use camino::{Utf8Path, Utf8PathBuf};
use chumsky::{input::Input as _, span::Span as _, Parser as _};
use codespan_reporting::files::SimpleFiles;
use diagnostics::error::{convert, Error};
use lasso::ThreadedRodeo;
use parser::ast::{Ast, Item};
use span::{FileId, Span};
use std::sync::LazyLock;

pub mod diagnostics;
pub mod lexer;
pub mod parser;
pub mod scopes;
pub mod span;

static RODEO: LazyLock<ThreadedRodeo> = LazyLock::new(ThreadedRodeo::new);

#[must_use]
pub fn parse_file(
    filename: &Utf8Path,
    files: &mut SimpleFiles<Utf8PathBuf, String>,
    errors: &mut Vec<Error>,
) -> Option<Ast> {
    let source = std::fs::read_to_string(filename).unwrap();
    let file_id = FileId::new(files.add(filename.to_owned(), source.clone())).unwrap();

    let (tokens, lexer_errors) = lexer::lexer()
        .parse(source.with_context(file_id))
        .into_output_errors();

    errors.extend(lexer_errors.iter().flat_map(|error| convert(error)));

    let (mut ast, parser_errors) = tokens.as_ref().map_or_else(
        || (None, vec![]),
        |tokens| {
            let eoi = tokens
                .last()
                .map_or_else(|| Span::zero(file_id), |(_, span)| span.to_end());

            parser::parser()
                .parse(tokens.spanned(eoi))
                .into_output_errors()
        },
    );

    errors.extend(parser_errors.iter().flat_map(|error| convert(error)));

    if let Some(ast) = ast.as_mut() {
        insert_explicit_submodules(ast, filename, files, errors);
    }

    ast
}

pub fn insert_explicit_submodules(
    ast: &mut Ast,
    ast_filename: &Utf8Path,
    files: &mut SimpleFiles<Utf8PathBuf, String>,
    errors: &mut Vec<Error>,
) {
    ast.0
        .iter_mut()
        .filter_map(|item| match &mut item.0 {
            Item::Module(submodule) => Some(submodule),
            _ => None,
        })
        .filter(|submodule| submodule.ast.is_none())
        .for_each(|submodule| {
            let filenames = {
                let directory = ast_filename
                    .parent()
                    .unwrap()
                    .join(submodule.name.resolve());

                let submodule_filename = format!("{}.pr", submodule.name.resolve());

                [
                    ast_filename.with_file_name(&submodule_filename),
                    directory.join("mod.pr"),
                    directory.join(submodule_filename),
                ]
            };

            let mut existing = filenames.iter().filter(|filename| filename.exists());

            let existing_cloned = existing.clone();

            let filename = existing.next().unwrap_or_else(|| todo!());

            if existing.next().is_some() {
                errors.push(Error::AmbiguousModule {
                    module_name: submodule.name.resolve(),
                    module_name_span: submodule.name.1,
                    filenames: existing_cloned.map(ToString::to_string).collect(),
                });

                return;
            }

            let ast = parse_file(filename, files, errors);

            submodule.ast = ast;
        });
}

#[cfg(test)]
mod tests {
    use crate::parse_file;
    use camino::Utf8PathBuf;
    use codespan_reporting::files::SimpleFiles;
    use owo_colors::OwoColorize as _;
    use std::{io::Write as _, path::PathBuf};

    #[test]
    fn parser_tests() {
        insta::glob!("../tests/parser", "**/*.pr", |path| {
            let path = Utf8PathBuf::from_path_buf(path.to_path_buf()).unwrap();

            let short_path = path
                .components()
                .skip_while(|c| c.as_os_str() != "tests")
                .collect::<PathBuf>();

            insta::elog!("{} {} ...", "parsing".cyan(), short_path.display());

            let mut files = SimpleFiles::new();
            let mut errors = vec![];
            let ast = parse_file(&path, &mut files, &mut errors);

            assert!(errors.is_empty(), "errors: {errors:#?}");

            let snapshot_suffix = short_path
                .components()
                .map(|c| c.as_os_str().to_string_lossy())
                .collect::<Vec<_>>()
                .join("_");

            insta::with_settings!({
                description => std::fs::read_to_string(path).unwrap().trim(),
                snapshot_suffix => snapshot_suffix,
            }, {
                insta::assert_yaml_snapshot!(ast.unwrap());
            });
        });
    }
}
