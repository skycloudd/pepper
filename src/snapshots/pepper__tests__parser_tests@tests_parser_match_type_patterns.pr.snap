---
source: src/lib.rs
description: "func type_patterns(aaaa: Aaaa, bbbb: Bbbb) -> Float32 =\n    var _: Float32 = match aaaa {\n        Aaaa { aa: 2.5, bb: 1.5 } => 1.5,\n        Aaaa { aa, bb } => n * m,\n    } in\n\n    match bbbb {\n        Bbbb.Cccc(2) => 1.5,\n        Bbbb.Cccc(n) => n * 3. / 2.,\n        Bbbb.Dddd => n * 4. / 2. + 1.,\n    }\n\nstruct Aaaa {\n    aa: Float32,\n    bb: Float32,\n}\n\nenum Bbbb {\n    Cccc(int32),\n    Dddd,\n}"
expression: ast.unwrap()
input_file: tests/parser/match/type_patterns.pr
snapshot_kind: text
---
- Function:
    name: type_patterns
    params:
      - name: aaaa
        ty:
          Primitive:
            base: Aaaa
            segments: []
      - name: bbbb
        ty:
          Primitive:
            base: Bbbb
            segments: []
    return_ty:
      Primitive:
        base: Float32
        segments: []
    body:
      Binding:
        name: _
        ty:
          Primitive:
            base: Float32
            segments: []
        value:
          Match:
            expr:
              Name:
                base: aaaa
                segments: []
            arms:
              - pattern:
                  pattern_type:
                    StructType:
                      name:
                        base: Aaaa
                        segments: []
                      fields:
                        - name: aa
                          pattern:
                            pattern_type:
                              Float: "2.5"
                            condition: ~
                        - name: bb
                          pattern:
                            pattern_type:
                              Float: "1.5"
                            condition: ~
                  condition: ~
                body:
                  Float: "1.5"
              - pattern:
                  pattern_type:
                    StructType:
                      name:
                        base: Aaaa
                        segments: []
                      fields:
                        - name: aa
                          pattern: ~
                        - name: bb
                          pattern: ~
                  condition: ~
                body:
                  BinaryOp:
                    op: Mul
                    lhs:
                      Name:
                        base: n
                        segments: []
                    rhs:
                      Name:
                        base: m
                        segments: []
        body:
          Match:
            expr:
              Name:
                base: bbbb
                segments: []
            arms:
              - pattern:
                  pattern_type:
                    TupleType:
                      name:
                        base: Bbbb
                        segments:
                          - Cccc
                      fields:
                        - pattern_type:
                            Int: "2"
                          condition: ~
                  condition: ~
                body:
                  Float: "1.5"
              - pattern:
                  pattern_type:
                    TupleType:
                      name:
                        base: Bbbb
                        segments:
                          - Cccc
                      fields:
                        - pattern_type:
                            Name:
                              base: n
                              segments: []
                          condition: ~
                  condition: ~
                body:
                  BinaryOp:
                    op: Div
                    lhs:
                      BinaryOp:
                        op: Mul
                        lhs:
                          Name:
                            base: n
                            segments: []
                        rhs:
                          Float: "3."
                    rhs:
                      Float: "2."
              - pattern:
                  pattern_type:
                    Name:
                      base: Bbbb
                      segments:
                        - Dddd
                  condition: ~
                body:
                  BinaryOp:
                    op: Add
                    lhs:
                      BinaryOp:
                        op: Div
                        lhs:
                          BinaryOp:
                            op: Mul
                            lhs:
                              Name:
                                base: n
                                segments: []
                            rhs:
                              Float: "4."
                        rhs:
                          Float: "2."
                    rhs:
                      Float: "1."
- Struct:
    name: Aaaa
    fields:
      - name: aa
        ty:
          Primitive:
            base: Float32
            segments: []
      - name: bb
        ty:
          Primitive:
            base: Float32
            segments: []
- Enum:
    name: Bbbb
    variants:
      - Tuple:
          name: Cccc
          fields:
            - Primitive:
                base: int32
                segments: []
      - Unit: Dddd
